src/custom-element.js

```diff
@@ -343,1087 +343,1100 @@ function isInternalOrServiceNode(node) {
+ * @param {!Window} win The window in which to register the custom element.
+ * @param {string} name The name of the custom element.
  * @param {function(new:./base-element.BaseElement, !Element)=} opt_implementationClass For
  *     testing only.
  * @return {!Object} Prototype of element.
  */
 export function createAmpElementProto(win, name, opt_implementationClass) {
-  const ElementProto = win.Object.create(createBaseAmpElementProto(win));
-  ElementProto.name = name;
+  const ElementProto = createBaseCustomElementClass(win, name).prototype;
   if (getMode().test && opt_implementationClass) {
     ElementProto.implementationClassForTesting = opt_implementationClass;
   }
   return ElementProto;
 }
 
 /**
- * Creates a new custom element class prototype.
- *
- * The prototype is cached per window and meant to be sub classed for a
- * concrete object.
+ * Creates a new custom element class.
  *
- * @param {!Window} win The window in which to register the elements.
- * @return {!Object} Prototype of element.
+ * @param {!Window} win The window in which to register the custom element.
+ * @param {string} name The name of the custom element.
+ * @return {Function} The custom element class.
  */
-function createBaseAmpElementProto(win) {
-
-  if (win.BaseCustomElementProto) {
-    return win.BaseCustomElementProto;
-  }
-
-  /**
-   * @lends {AmpElement.prototype}
-   */
-  const ElementProto = win.Object.create(win.HTMLElement.prototype);
-  win.BaseCustomElementProto = ElementProto;
-
-  /**
-   * Called when elements is created. Sets instance vars since there is no
-   * constructor.
-   * @final @this {!Element}
-   */
-  ElementProto.createdCallback = function() {
-    // Flag "notbuilt" is removed by Resource manager when the resource is
-    // considered to be built. See "setBuilt" method.
-    /** @private {boolean} */
-    this.built_ = false;
-
-    /** @type {string} */
-    this.readyState = 'loading';
-
-    /** @type {boolean} */
-    this.everAttached = false;
-
+function createBaseCustomElementClass(win, name) {
+  const superclass = /** @type {function(new:HTMLElement)} */ (win.HTMLElement);
+  return class BaseCustomElement extends superclass {
     /**
-     * Ampdoc can only be looked up when an element is attached.
-     * @private {?./service/ampdoc-impl.AmpDoc}
+     * Constructor uses atypical pattern as part of a workaround for ES6
+     * transpilation issue when extending HTMLElement.
+     * @see https://github.com/WebReflection/document-register-element#v1-caveat
+     * @suppress {checkTypes}
      */
-    this.ampdoc_ = null;
+    constructor(self) {
+      self = super(self);
+      self.init();
+      return self;
+    }
 
     /**
-     * Resources can only be looked up when an element is attached.
-     * @private {?./service/resources-impl.Resources}
+     * Sets up properties for this class.
+     * @note Suppresses type check for Custom Elements V0 compatibility.
+     * @suppress {checkTypes}
```

Comment:
we should be able to remove this suppress once the whole codebase is type checked right? if so can we add a TODO to remove this